File: polars_queries/q1.py
from datetime import datetime

import polars as pl

from polars_queries import utils

Q_NUM = 1


def q():
    var_1 = datetime(1998, 9, 2)
    q = utils.get_line_item_ds()
    q_final = (
        q.filter(pl.col("l_shipdate") <= var_1)
        .group_by(["l_returnflag", "l_linestatus"])
        .agg(
            [
                pl.sum("l_quantity").alias("sum_qty"),
                pl.sum("l_extendedprice").alias("sum_base_price"),
                (pl.col("l_extendedprice") * (1 - pl.col("l_discount")))
                .sum()
                .alias("sum_disc_price"),
                (
                    pl.col("l_extendedprice")
                    * (1.0 - pl.col("l_discount"))
                    * (1.0 + pl.col("l_tax"))
                )
                .sum()
                .alias("sum_charge"),
                pl.mean("l_quantity").alias("avg_qty"),
                pl.mean("l_extendedprice").alias("avg_price"),
                pl.mean("l_discount").alias("avg_disc"),
                pl.count().alias("count_order"),
            ],
        )
        .sort(["l_returnflag", "l_linestatus"])
    )

    utils.run_query(Q_NUM, q_final)


if __name__ == "__main__":
    q()

File: misc/polars_queries_rust/src/q1.rs
use crate::utils::*;
use chrono::NaiveDate;
use polars::prelude::*;

pub fn query() -> PolarsResult<LazyFrame> {
    let q = get_lineitem_ds();
    let var_1 = lit(NaiveDate::from_ymd_opt(1998, 9, 2)
        .unwrap()
        .and_hms_opt(0, 0, 0)
        .unwrap());

    let q = q
        .filter(col("l_shipdate").lt_eq(var_1))
        .groupby([cols(["l_returnflag", "l_linestatus"])])
        .agg([
            sum("l_quantity").alias("sum_qty"),
            sum("l_extendedprice").alias("sum_base_price"),
            (col("l_extendedprice") * (lit(1) - col("l_discount")))
                .sum()
                .alias("sum_disc_price"),
            (col("l_extendedprice") * (lit(1.0) - col("l_discount")) * (lit(1.0) + col("l_tax")))
                .sum()
                .alias("sum_charge"),
            mean("l_quantity").alias("avg_qty"),
            mean("l_extendedprice").alias("avg_price"),
            mean("l_discount").alias("avg_disc"),
            count().alias("count_order"),
        ])
        .sort_by_exprs([cols(["l_returnflag", "l_linestatus"])], &[false], false);

    Ok(q)
}
